Extend create-vm to allow targeting specific repositories / channels

KSerrania opened the pull request:
What does this PR do?
---------------------

Adds two new options to the `invoke create-vm` invoke task:
- `--agent-repository`, which allows choosing the repository to use (between `prod`, `staging`, `testing`, and `trial`)
- `--agent-channel`, which allows choosing the distribution channel within a repository.

Some other changes needed to achieve this:
- modification of the AgentVersion struct to contain the repository and channel data
- use of the non-deprecated Linux install script every time (to support `DD_REPO_URL`, the preferred repository option)
- handle repositories with non-standard paths, which need to be taken into account (prod/stable, and staging, on Windows).

Which scenarios this will impact?
-------------------

Any scenario that installs the Agent on a host.

Motivation
----------

Being able to install any published Agent version, not only the ones in prod/stable and staging/beta.

Additional Notes
----------------

Future work:
- Support for Windows testing repositories. The easiest way to get this is likely to make the `datadog-agent` pipeline create an `installers_v2.json` file in the testing repositories.
- Clean support for targeting the "latest Agent 6" version. Right now, there is no official way to target a given major version of the Agent without specifying the minor (you can by passing `--agent-version "6."`, but it's due to an implementation detail).


A few questions / notes:
- should I add a check to make the `invoke create-vm` task fail early if the specified Agent repository is not in the supported list?
- for simplicity, the Agent repository variable is a shorthand for the actual repositories (eg. `staging` -> `apt.datad0g.com` / `yum.datad0g.com` / `dd-agent-mstesting.s3.amazonaws.com/builds`). Should we also allow letting users specify arbitrary URLs?
- I prefer making all Agent-related options prefixed with `agent`, so I renamed `--pipeline-id` into `--agent-pipeline-id`.

Testing
----------------

Run:
`invoke create-vm -o ubuntu`
`invoke create-vm -o redhat`
`invoke create-vm -o suse`
`invoke create-vm -o windows`

`invoke create-vm --agent-repository prod --agent-channel beta -o ubuntu`
`invoke create-vm --agent-repository prod --agent-channel beta -o redhat`
`invoke create-vm --agent-repository prod --agent-channel beta -o suse`
`invoke create-vm --agent-repository prod --agent-channel beta -o windows`

`invoke create-vm --agent-repository staging --agent-channel stable -o ubuntu`
`invoke create-vm --agent-repository staging --agent-channel stable -o redhat`
`invoke create-vm --agent-repository staging --agent-channel stable -o suse`
`invoke create-vm --agent-repository staging --agent-channel stable -o windows`

`invoke create-vm --agent-repository staging --agent-channel beta -o ubuntu`
`invoke create-vm --agent-repository staging --agent-channel beta -o redhat`
`invoke create-vm --agent-repository staging --agent-channel beta -o suse`
`invoke create-vm --agent-repository staging --agent-channel beta -o windows`

`invoke create-vm --agent-repository staging --agent-channel nightly -o ubuntu`
`invoke create-vm --agent-repository staging --agent-channel nightly -o redhat`
`invoke create-vm --agent-repository staging --agent-channel nightly -o suse`
`invoke create-vm --agent-repository staging --agent-channel nightly -o windows`

`invoke create-vm --agent-repository trial --agent-channel stable -o ubuntu`
`invoke create-vm --agent-repository trial --agent-channel stable -o redhat`
`invoke create-vm --agent-repository trial --agent-channel stable -o suse`
`invoke create-vm --agent-repository trial --agent-channel stable -o windows` # Will fail because we haven't generated installers_v2.json files Windows trial repositories yet

`invoke create-vm --agent-repository trial --agent-channel nightly -o ubuntu`
`invoke create-vm --agent-repository trial --agent-channel nightly -o redhat`
`invoke create-vm --agent-repository trial --agent-channel nightly -o suse`
`invoke create-vm --agent-repository trial --agent-channel nightly -o windows` # Will fail because we haven't generated installers_v2.json files Windows trial repositories yet

`invoke create-vm --agent-pipeline-id <valid pipeline id> -o ubuntu`
`invoke create-vm --agent-pipeline-id <valid pipeline id> -o redhat`
`invoke create-vm --agent-pipeline-id <valid pipeline id>  -o suse`
`invoke create-vm --agent-pipeline-id <valid pipeline id> -o windows` # Should fail because Windows testing repositories don't have installers_v2.json files

Check that all of these succeed (except otherwise mentioned), and verify their Agent version.

vboulineau commented:
It seems to be conflicting with https://github.com/DataDog/test-infra-definitions/pull/127/files

KSerrania commented:
> 
It seems to be conflicting with https://github.com/DataDog/test-infra-definitions/pull/127/files

Indeed, as this is an extension of the functionality it adds. I can wait until it's merged and then rebase my PR.

KSerrania commented:
Rebased on top of https://github.com/DataDog/test-infra-definitions/pull/127/files and changed the base branch to the PR to make the diff easier to parse.

I'll double-check that nothing broke after the rebase.
